src/coreclr/jit/hwintrinsiccodegenarm64.cpp

```diff
@@ -489,6 +504,40 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
 
         switch (intrin.id)
         {
+            case NI_AdvSimd_VectorTableLookup_2:
+            case NI_AdvSimd_Arm64_VectorTableLookup_2:
+            case NI_AdvSimd_VectorTableLookup_3:
+            case NI_AdvSimd_Arm64_VectorTableLookup_3:
+            case NI_AdvSimd_VectorTableLookup_4:
+            case NI_AdvSimd_Arm64_VectorTableLookup_4:
+            {
+
+                if (!intrin.op1->OperIsFieldList())
+                {
+                    assert(!"Expect the first operand of VectorTableLookup to be FIELD_LIST");
+                }
+
+                GenTreeFieldList* fieldList  = intrin.op1->AsFieldList();
+                GenTree*          firstField = fieldList->Uses().GetHead()->GetNode();
+
+                if (firstField->IsCopyOrReload())
+                {
+                    // If value is copied in a register to satisfy the consecutive-register
+                    // requirement, make sure to get the source's register because these
+                    // instruction encoding takes only the 1st register and infer the rest
+                    // from that.
+                    GenTree* op1 = firstField->AsCopyOrReload()->gtGetOp1();
+                    assert(!op1->IsCopyOrReload());
+                    op1Reg = op1->GetRegNum();
```

Comment:
I am not sure I understand how this works. What guarantees the original register holds a valid value at this point (or does it not matter, or...)?

Reply:
I am not sure if I follow. We create `RefPosition` based on the `GenTree` nodes, allocate registers to them and then write back the register assignments back to the tree nodes. If the register assigned at given point is different than what the original register it was assigned, we have a copy node, which we use to save the value in appropriate register at that point, before its use. This works just as any other place in codegen.

Reply:
What I do not understand here is why getting the register from the source of copy/reload is correct.

Let's consider the reload case. We have `firstField` assigned `REG_0` at def, it is spilled and the use has assigned `REG_1` to it. This should be:
```
t1 =  FIRST_FIELD_NODE REG_0
t2  = GT_RELOAD(t1) REG_1
      OUR_INTRINSIC(t2)
```
And here the code seemingly uses `REG_0` for `OUR_INTRINSIC`, what guarantess it will have the same value as `REG_1` that was loaded from the spill temp?

Reply:
You are right, it doesn't. After my other changes, I don't think we need to handle the case for `CopyOrReload`. It should just use the node's register. I have also added an assert to make sure the subsequent refpositions gets consecutive registers.